sql/wsrep_xid.cc

```diff
@@ -20,6 +20,44 @@
 #include "sql_class.h"
 #include "wsrep_mysqld.h" // for logging macros
 
+static inline wsrep_seqno_t
+wsrep_seqno_byteswap(wsrep_seqno_t const seqno)
+{
+  union { wsrep_seqno_t s; uint8_t b[8]; } from, to;
+
+  from.s = seqno;
+
+  to.b[0] = from.b[7]; to.b[1] = from.b[6];
+  to.b[2] = from.b[5]; to.b[3] = from.b[4];
+  to.b[4] = from.b[3]; to.b[5] = from.b[2];
+  to.b[6] = from.b[1]; to.b[7] = from.b[0];
+
+  return to.s;
+}
+
+/* Since vast majority of existing installations are little-endian and
+ * for the general little-endian cause, canonical XID seqno representation
+ * is little-endian. */
+#ifdef WORDS_BIGENDIAN
+  #define host_to_xid_seqno(s) wsrep_seqno_byteswap(s)
+  static inline wsrep_seqno_t
+  xid_to_host_seqno(wsrep_seqno_t const seqno)
+  {
+    // a little trick to simplify migration for existing installations:
+    // it is highly inlikely that the current seqno in any big-endian
+    // production cluster exceeds 0x0000000fffffffffLL (64G).
+    // Hence any seqno read from XID that is bigger than that must have been
+    // written in BE format. This BE detection has failure probability of
+    // 1 in 256M. This must be removed after next release.
+    wsrep_seqno_t const ret(wsrep_seqno_byteswap(seqno));
+    return (ret > 0x0000000fffffffffLL || ret < WSREP_SEQNO_UNDEFINED ?
```

Comment:
This looks like a time bomb. It would probably be better to change WSREP_XID_PREFIX and do the byteswap if the stored value has the new format.